README.md


[![PyPI Latest Release](https://img.shields.io/pypi/v/PriceIndices.svg)](https://pypi.org/project/PriceIndices/)
[![Coverage](https://codecov.io/github/dc-aichara/PriceIndices/coverage.svg?branch=master)](https://codecov.io/gh/dc-aichara/PriceIndices)
[![License](https://img.shields.io/pypi/l/PriceIndices.svg)](https://github.com/dc-aichara/PriceIndices/blob/master/LICENSE)
[![Downloads](https://static.pepy.tech/personalized-badge/priceindices?period=month&units=international_system&left_color=gray&right_color=blue&left_text=PyPI%20downloads%20per%20month)](https://pepy.tech/project/priceindices)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![Imports: isort](https://img.shields.io/badge/%20imports-isort-%231674b1?style=flat&labelColor=ef8336)](https://pycqa.github.io/isort/)

## Development Environment

### Poetry
 * Install Poetry
    ```shell
    curl -sSL https://install.python-poetry.org | python3 -

    ```
 * Install dependencies
    ```shell
    poetry install
    ```
 * To add new dependencies use `poetry add` 
    ```shell
    poetry add dependency_name
    ```
 * Read `Poetry` [documentation](https://python-poetry.org/docs/) for more.

## Installation 

### pip 

```
pip install PriceIndics
```

### Poetry

```
poetry add PriceIndices

```

### From Source (Github)
 
 git clone https://github.com/dc-aichara/Price-Indices.git
 
 cd Price-Indices 
 
 python3 setup.py install
 
## Usages 
 
```python
from PriceIndices import MarketHistory, Indices

```
## Examples 

- #### Get market history and closing price

```python
>>> history = MarketHistory()

# Get Market History 

>>> df_history = history.get_history("bitcoin", "2020-03-16", "2021-03-15")  
>>> df_history.head()
         open          high           low         close        volume    market_cap        date
0  59267.429049  60540.992712  55393.165363  55907.200226  6.641937e+10  1.042946e+12  2021-03-15
1  61221.134297  61597.918396  59302.316977  59302.316977  4.390123e+10  1.106226e+12  2021-03-14
2  57343.370247  61683.864014  56217.972382  61243.084766  6.066983e+10  1.142369e+12  2021-03-13
3  57821.218747  57996.619490  55376.650088  57332.088964  5.568994e+10  1.069366e+12  2021-03-12
4  55963.180089  58091.062703  54484.593089  57805.123019  5.677234e+10  1.078136e+12  2021-03-11

# Get closing price

>>> price_data  =  history.get_price("bitcoin", "2020-03-16", "2021-03-15") 

>>> price_data.head()
         date         price
0  2021-03-15  55907.200226
1  2021-03-14  59302.316977
2  2021-03-13  61243.084766
3  2021-03-12  57332.088964
4  2021-03-11  57805.123019

```

- #### Calculate Volatility Index

```python
indices = Indices(df=price_data, plot_dir="plots")
>>> df_bvol = indices.get_vola_index(
        plot=True,
        plot_name="vola_index.png",
        show_plot=False  
)  
>>> df_bvol.head()
        date    price  BVOL_Index
0 2019-10-29  9427.69    0.711107
1 2019-10-28  9256.15    0.707269
2 2019-10-27  9551.71    0.709765
3 2019-10-26  9244.97    0.698544
4 2019-10-25  8660.70    0.692656

```

- #### Plot Volatility Index
 Plot will be saved in plots directory as `vola_index.png`.


<img src= 'plots/bvol_index.png' >

- #### Calculate Relative Strength Index (RSI)

```python

>>> df_rsi = indices.get_rsi(
        plot=True,
        plot_name="rsi.png",
        show_plot=False,
)   

>>> print(df_rsi.head())
        date    price       RSI_1  RS_Smooth      RSI_2
0 2019-10-30  9205.73      64.641855   1.624958  61.904151
1 2019-10-29  9427.69      65.707097   1.709072  63.086984
2 2019-10-28  9256.15      61.333433   1.597755  61.505224
3 2019-10-27  9551.71      66.873327   2.012345  66.803267
4 2019-10-26  9244.97      63.535368   1.791208  64.173219


```

- #### Plot RSI
Plot will be saved in plots directory as `rsi.png`.


<img src='plots/rsi.png' >

- #### Get Bollinger Bands and its plot

```python
>>> df_bb = indices.get_bollinger_bands(
        days=20, 
        plot=True,
        plot_name="bollinger_bands.png",
        show_plot=False,
        ) 
>>> df_bb.head()
        date    price     BB_upper   BB_lower
0 2019-10-30  9205.73  9635.043581 -8428.5855
1 2019-10-29  9427.69  9550.707153 -8397.6225
2 2019-10-28  9256.15  9408.263164 -8356.0250
3 2019-10-27  9551.71  9268.466516 -8304.6565
4 2019-10-26  9244.97  9003.752779 -8239.3520


"""
This will also save Bollingers bands plot in your working directory as 'bollinger_bands.png' in plots folder.
"""

```

<img src='plots/bollinger_bands.png'>


- #### Get Moving Average Convergence Divergence (MACD) and its plot

```python

>>> df_macd = indices.get_moving_average_convergence_divergence(
        plot=True,
        plot_name="macd.png",
        show_plot=False,
)
"""
This will return a pandas DataFrame and save EMA plot as 'macd.png' in in plots folder. 
""""
>>> df_macd.head()
        date    price       MACD
0 2019-10-30  9205.73   0.000000
1 2019-10-29  9427.69  17.706211
2 2019-10-28  9256.15  17.692715
3 2019-10-27  9551.71  41.057952
4 2019-10-26  9244.97  34.426864


```

<img src='plots/macd.png'>

- #### Get Simple Moving Average (SMA) and its plot

```python
>>> df_sma = indices.get_simple_moving_average(
        days=20,
        plot=True,
        plot_name="sma.png",
        show_plot=False,
) 
"""This will return a pandas DataFrame and save EMA plot as 'sma.png' in plots folder.
""""
>>> df_sma.head()
        date    price          SMA
0 2019-10-30  9205.73  8467.488000
1 2019-10-29  9427.69  8400.797333
2 2019-10-28  9256.15  8330.597333
3 2019-10-27  9551.71  8268.254667
4 2019-10-26  9244.97  8187.244667


```

<img src='plots/sma.png'>

- ### Get Exponential Moving Average (EMA) and its plot

```python
>>> df_ema = indices.get_exponential_moving_average(
        periods=(20,70),
        plot=True,
        plot_name="ema.png",
        show_plot=False,
)
"""This will return a pandas DataFrame and save EMA plot as 'ema.png' in plots folder.
""""

>>> df_ema.head()
        date    price       EMA_20       EMA_70
0 2019-10-30  9205.73  9205.730000  9205.730000
1 2019-10-29  9427.69  9226.869048  9211.982394
2 2019-10-28  9256.15  9229.657710  9213.226552
3 2019-10-27  9551.71  9260.329356  9222.761297
4 2019-10-26  9244.97  9258.866561  9223.386895
>>> 


```

<img src='plots/ema.png' >

### License
 
[MIT](https://choosealicense.com/licenses/mit/) © [Dayal Chand Aichara](https://github.com/dc-aichara)


### Check out [webpage](https://dc-aichara.github.io/PriceIndices/) of PriceIndices package. 

### I have created a [cryptocurrency technical indicators dashboard](https://crypto-indicators-dashboard.herokuapp.com/) which uses this library. 
### Disclaimer: 

```
All content provided here, is for educational purpose and your general information only, procured  from third party sources.
I make no warranties of any kind in relation to this content, including but  not limited to accuracy
and updatedness. No part of the content that I provide  constitutes  financial  advice, legal advice 
or any other form of advice meant for your specific reliance for any purpose. Any use or reliance on
my content is solely at your own risk  and  discretion. You should conduct your own research, review, 
analyse and  verify my content  before relying  on them. Trading is a highly risky activity that can 
lead to  major  losses, please  therefore  consult your financial advisor before making any decision.
No content on this Site is meant to be a solicitation or offer.
```


ChangeLog.md
# PriceIndices ChangeLog

## Version 

## 1.4.0

* Refactored code
* Adapted poetry for development
* Updated documentation.

## Version 1.3.0

* Optimize Indicators calculations.
* Minor bug fixes.
* Refactored code.

## Version 1.2.1

- Fixed documentation.

## Version 1.2.0

- Added function annotations
- Added `__init__` to Indices class. Indices can should be initiated with  a Pandas DataFrame, date column (optional) , and price column (optional).
- Function name `get_bvol_index`, and `get_bvol_graph` changed to `get_vola_index`, and `get_vola_graph` respectively.

- Added plot output input to plot functions.  

tests/test_price_indices.py
from PriceIndices import Indices, MarketHistory

history = MarketHistory()

price_data = history.get_price("bitcoin", "2020-03-16", "2021-03-15")
indices = Indices(df=price_data, plot_dir="tests/plots")


def test_vola():
    df_bvol = indices.get_vola_index(
        plot=True, plot_name="vola_index.png", show_plot=False
    )

    assert df_bvol.shape == (334, 3)


def test_rsi():
    df_rsi = indices.get_rsi(
        plot=True,
        plot_name="rsi.png",
        show_plot=False,
    )
    assert df_rsi.shape == (363, 5)


def test_bollinger_bands():
    df_bb = indices.get_bollinger_bands(
        days=20,
        plot=True,
        plot_name="bollinger_bands.png",
        show_plot=False,
    )

    assert df_bb.shape == (364, 4)


def test_moving_average_convergence_divergence():
    df_macd = indices.get_moving_average_convergence_divergence(
        plot=True,
        plot_name="macd.png",
        show_plot=False,
    )
    assert df_macd.shape == (364, 3)


def test_simple_moving_average():
    df_sma = indices.get_simple_moving_average(
        days=20,
        plot=True,
        plot_name="sma.png",
        show_plot=False,
    )

    assert df_sma.shape == (345, 3)


def test_exponential_moving_average():
    df_ema = indices.get_exponential_moving_average(
        periods=(20, 70),
        plot=True,
        plot_name="ema.png",
        show_plot=False,
    )

    assert df_ema.shape == (364, 4)


def test_if_plots_saved():
    for plot_name in [
        "bollinger_bands",
        "sma",
        "vola_index",
        "ema",
        "macd",
        "rsi",
    ]:
        full_plot_path = indices.plot_dir.joinpath(f"{plot_name}.png")
        assert full_plot_path.exists() == 1
        # Clean plots folder: Remove all saved plot
        full_plot_path.unlink()


tests/test_data_extraction.py
from PriceIndices import MarketHistory


def test_history_market_data():
    history = MarketHistory()
    df_history = history.get_history("bitcoin", "2020-03-16", "2021-03-15")
    assert df_history.shape == (364, 7)

    price_data = history.get_price("bitcoin", "2020-03-16", "2021-03-15")

    assert price_data.shape == (364, 2)


PriceIndices/crypto_history.py
import warnings
from typing import Optional

import pandas as pd
import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry

warnings.filterwarnings("ignore")


class MarketHistory(object):
    __Crypto_Market_Base_URL = "https://web-api.coinmarketcap.com/v1/cryptocurrency/ohlcv/historical?convert=USD&slug="  # noqa

    def __init__(
        self, base_url: Optional[str] = __Crypto_Market_Base_URL
    ) -> None:
        self.base_url = base_url
        self.request_timeout = 120

        self.session = requests.Session()
        retries = Retry(
            total=5, backoff_factor=0.5, status_forcelist=[502, 503, 504]
        )
        self.session.mount("http://", HTTPAdapter(max_retries=retries))

    def __request(self, url: str) -> Optional[pd.DataFrame]:
        try:
            response = self.session.get(url, timeout=self.request_timeout)
            response.raise_for_status()
            content = response.json()

            d = content["data"]["quotes"]
            df = pd.DataFrame([v["quote"]["USD"] for v in d])
            df.sort_values("timestamp", ascending=False, inplace=True)
            df.reset_index(drop=True, inplace=True)
            df["timestamp"] = pd.to_datetime(df["timestamp"])
            df["date"] = df["timestamp"].apply(lambda x: x.strftime("%Y-%m-%d"))

            del df["timestamp"]
            return df
        except Exception as e:
            raise e

    def get_history(
        self, coin_id: str, start_date: str, end_date: str
    ) -> Optional[pd.DataFrame]:
        """
        Get historical market data of a cryptocurrency from CoinMarketCap.
        Args:
            coin_id (str): coin name. E.g., bitcoin
            start_date (str): Starting date in 'YYYY-MM-DD' format
            end_date (str): End date in 'YYYY-MM-DD' format
        Returns:
            pd.DataFrame: Pandas Dataframe or print error message

        """
        url = "{0}{1}&time_end={2}&time_start={3}".format(
            self.base_url, coin_id, end_date, start_date
        )

        try:
            return self.__request(url)
        except Exception as e:
            print(e)
            print(
                "Please, check inputs. Coin id, and dates are strings. Date "
                'format is "YYYY-MM-DD"'
            )
            return None

    def get_price(
        self, coin_id: str, start_date: str, end_date: str
    ) -> Optional[pd.DataFrame]:
        """
        Get historical market price data (closing price) of a cryptocurrency
        from CoinMarketCap.
        Args:
            coin_id (str): coin name
            start_date (str): Starting date in 'YYYY-MM-DD' format
            end_date (str): End date in 'YYYY-MM-DD' format

        Returns:
            pd.DataFrame: Pandas Dataframe or print error message

        """
        url = "{0}{1}&time_end={2}&time_start={3}".format(
            self.base_url, coin_id, end_date, start_date
        )

        try:
            df = self.__request(url)
            assert type(df) == pd.DataFrame
            df = df[["date", "close"]]
            df.columns = ["date", "price"]
            return df
        except Exception as e:
            print(
                e,
                "Please, check inputs Coin id, and dates are strings. Date "
                'format is "YYYY-MM-DD"',
            )
            return None


PriceIndices/price_indicators.py
import warnings
from pathlib import Path
from typing import List, Optional

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

warnings.filterwarnings("ignore")


class Indices:
    """
    Price Technical Indicators
    """

    def __init__(
        self,
        df: pd.DataFrame,
        date_col: str = "date",
        price_col: str = "price",
        plot_dir: Optional[str] = "",
    ) -> None:
        self.df = df
        self.date_col = date_col
        self.price_col = price_col
        self.plot_dir = Path(plot_dir)  # type: ignore

    def get_vola_index(
        self,
        volatile_period: Optional[int] = 30,
        plot: Optional[bool] = False,
        plot_name: Optional[str] = "vola_index.png",
        show_plot: Optional[bool] = False,
    ) -> pd.DataFrame:
        """
        Volatility Index is a measure of market's expectation of volatility
        over the near term.
        Volatility is often described as the "rate and magnitude of changes in
        prices" and in finance often referred to as risk.
        Reference:
            www.moneycontrol.com
        Args:
            volatile_period (int): Period to calculate volatile index.
                                   Defulat to 30.
            plot (bool): If plot bollinger bands
            plot_name(str): Plot with ".png" extension.
            show_plot (bool): True if you plot should be displayed.

        Returns:
            pd.DataFrame: Pandas DataFrame
        """
        data = self.df.copy()
        data = data.sort_values(by=self.date_col).reset_index(drop=True)
        v = np.log(data[self.price_col]).diff().rolling(
            volatile_period
        ).std() * np.sqrt(365)
        df_bvol = pd.DataFrame(data={"BVOL_Index": v})
        data = pd.concat([data, df_bvol], join="inner", axis=1)
        data = data.dropna()
        data = data.sort_values(by=self.date_col, ascending=False).reset_index(
            drop=True
        )
        if plot is True:
            self.get_vola_graph(data, plot_name, show_plot)
        return data

    def get_vola_graph(
        self,
        data: pd.DataFrame,
        plot_name: Optional[str] = "bvol_index.png",
        show_plot: Optional[bool] = False,
    ) -> None:
        """
        Make a line graph of volatile index with respect to time
        Args:
            data(pd.DataFrame): Output of get_vola_index function
            plot_name(str): Plot with ".png" extension.
            show_plot: True if you plot should be displayed.
        """

        fig, ax = plt.subplots(figsize=(14, 12))
        rect = fig.patch
        rect.set_facecolor("yellow")
        ax1 = plt.subplot(211)
        ax1.plot(data["date"], data["price"], color="blue", label="Price")
        plt.ylabel("Price", color="red", fontsize=20)
        ax1.axes.get_xaxis().set_ticks([])
        plt.legend()
        ax1.tick_params(axis="y", colors="b")
        ax1.grid(color="grey", linestyle="-", linewidth=0.25, alpha=0.5)

        ax2 = plt.subplot(212)
        ax2.plot(
            data["date"], data["BVOL_Index"], color="b", label="BVOL Index"
        )
        plt.xlabel("Time", color="red", fontsize=20)
        plt.ylabel("Volatility Index", color="r", fontsize=20)
        plt.legend()
        plt.setp(ax2.xaxis.get_majorticklabels(), rotation=90)
        ax2.grid(color="grey", linestyle="-", linewidth=0.25, alpha=0.5)

        ax2.tick_params(axis="x", colors="b")
        ax2.tick_params(axis="y", colors="b")

        plt.suptitle("Price  and  Volatility Index", color="red", fontsize=24)
        plt.savefig(
            self.plot_dir.joinpath(plot_name),  # type: ignore
            bbox_inches="tight",
            facecolor="orange",
        )
        if show_plot:
            plt.show()

    def get_rsi(
        self,
        plot: Optional[bool] = False,
        plot_name: Optional[str] = "rsi.png",
        show_plot: Optional[bool] = False,
    ) -> pd.DataFrame:
        """
        Type:
            Momentum indicator

        Computation:
                    It is based on the average price increase during a period of
                    rising prices and average price fall during a period of
                    falling stock prices. Relative Strength Index (RSI) is
                    plotted between 0 and 100.

        What it signals:
                        Usually, the market is treated as overbought when RSI
                        goes above 70 (80 for highly volatile stocks) and
                        oversold when it hits 30—20 for highly volatile stocks.

        Reference:
                    https://economictimes.indiatimes.com/

        Args:
            plot (bool): If plot bollinger bands
            plot_name(str): Plot with ".png" extension.
            show_plot (bool): True if you plot should be displayed.

        Returns:
            pd.DataFrame: Pandas DataFrame with RSI values

        """
        data = self.df.copy()
        data = data.sort_values(by=self.date_col).reset_index(drop=True)
        data["price_change"] = data[self.price_col] - data[
            self.price_col
        ].shift(1)
        data.dropna(inplace=True)
        data["gain"] = np.where(
            data["price_change"] >= 0, data["price_change"], 0
        )
        data["loss"] = np.where(
            data["price_change"] <= 0, abs(data["price_change"]), 0
        )

        data["gain_average"] = data["gain"].rolling(14).mean()
        data["loss_average"] = data["loss"].rolling(14).mean()

        data["RS"] = data["gain_average"] / data["loss_average"]

        data["RSI_1"] = 100 * (1 - (1 / (1 + data["RS"])))

        data["RS_Smooth"] = (
            data["gain_average"].shift(1) * 13 + data["gain"]
        ) / (data["loss_average"].shift(1) * 13 + data["loss"])

        data["RSI_2"] = 100 * (1 - (1 / (1 + data["RS_Smooth"])))
        data = data.fillna(0).reset_index(drop=True)
        data.drop(
            [
                "gain",
                "loss",
                "price_change",
                "gain_average",
                "loss_average",
                "RS",
            ],
            axis=1,
            inplace=True,
        )
        data = data.sort_values(by=self.date_col, ascending=False).reset_index(
            drop=True
        )
        if plot is True:
            self.get_rsi_graph(data, plot_name, show_plot)
        return data

    def get_rsi_graph(
        self,
        data: pd.DataFrame,
        plot_name: Optional[str] = "rsi.png",
        show_plot: Optional[bool] = False,
    ) -> None:
        """
        Plot RSI against date and price
        Args:
            data(pd.DataFrame): Output of get_rsi function.
            plot_name(str): Plot with ".png" extension.
            show_plot: True if you plot should be displayed.


        """
        fig, ax = plt.subplots(figsize=(14, 12))
        rect = fig.patch
        rect.set_facecolor("yellow")
        ax1 = plt.subplot(211)
        ax1.plot(data["date"], data["price"], color="blue", label="Price")
        plt.ylabel("Price ($)", color="red", fontsize=20)
        ax1.axes.get_xaxis().set_ticks([])
        plt.legend()
        ax1.tick_params(axis="y", colors="b")

        ax2 = plt.subplot(212)
        ax2.plot(data["date"], data["RSI_2"], color="b", label="RSI")
        plt.xlabel("Time", color="red", fontsize=20)
        plt.ylabel("Relative Strength Index (RSI)", color="r", fontsize=20)
        plt.text(
            data["date"][int(len(data) / 2)],
            80,
            ">70 OverBought",
            fontsize=20,
            color="black",
        )
        plt.text(
            data["date"][int(len(data) / 2)],
            15,
            "<30 OverSold",
            fontsize=20,
            color="black",
        )
        plt.legend()
        plt.setp(ax2.xaxis.get_majorticklabels(), rotation=90)

        ax2.tick_params(axis="x", colors="b")
        ax2.tick_params(axis="y", colors="b")

        ax2.axhline(y=70, color="r")
        ax2.axhline(y=30, color="r")

        plt.suptitle(
            "Price  and  Relative  Strength Index", color="red", fontsize=24
        )
        plt.savefig(
            self.plot_dir.joinpath(plot_name),  # type: ignore
            bbox_inches="tight",
            facecolor="orange",
        )
        if show_plot:
            plt.show()

    def get_bollinger_bands(
        self,
        days: Optional[int] = 20,
        plot: Optional[bool] = False,
        plot_name: Optional[str] = "bollinger_bands.png",
        show_plot: Optional[bool] = False,
    ) -> pd.DataFrame:
        """
        Type:
            Trend, volatility, momentum indicator

        Computation:
                    They comprise three lines: A 20-day moving average, an upper
                    band and lower band—the upper and lower bands are plotted as
                    two standard deviations from the moving average.

        What it signals:
                        The moving average shows the trend, the gap between
                        upper and lower band shows volatility in the counter.

        References:
                   1.  https://economictimes.indiatimes.com/
                   2. https://www.bollingerbands.com/bollinger-bands
        Args:
            days (int): Number of days to calculate moving average
            plot (bool): If plot bollinger bands
            plot_name(str): Plot with ".png" extension.
            show_plot: True if you plot should be displayed.

        Returns:
            pd.DataFrame: A pandas DataFrame and save a plot to given path.

        """
        data = self.df.copy()
        data = data.sort_values(by=self.date_col).reset_index(drop=True)
        data["SMA"] = data[self.price_col].rolling(days).mean()
        data["SD"] = data[self.price_col].rolling(days).std()
        data["BB_upper"] = data["SMA"] + data["SD"] * 2
        data["BB_lower"] = data["SMA"] - data["SMA"] * 2
        data.drop(["SD", "SMA"], axis=1, inplace=True)
        data = data.sort_values(by=self.date_col, ascending=False).reset_index(
            drop=True
        )
        if plot:
            fig, ax = plt.subplots(figsize=(16, 12))
            plt.plot(data[self.date_col], data["BB_upper"], color="g")
            plt.plot(data[self.date_col], data["BB_lower"], color="g")
            plt.plot(data[self.date_col], data[self.price_col], color="orange")
            plt.legend()
            plt.xlabel("Time", color="b", fontsize=22)
            plt.ylabel("Price", color="b", fontsize=22)
            plt.title("Bollinger Bands", color="b", fontsize=27)
            plt.tick_params(labelsize=17)
            fig.set_facecolor("yellow")
            plt.grid()
            plt.savefig(
                self.plot_dir.joinpath(plot_name),  # type: ignore
                bbox_inches="tight",
                facecolor="orange",
            )
            if show_plot:
                plt.show()
        return data

    def get_moving_average_convergence_divergence(
        self,
        plot: Optional[bool] = False,
        plot_name: Optional[str] = "macd.png",
        show_plot: Optional[bool] = False,
    ) -> pd.DataFrame:
        """
        Type
            Trend and momentum indicator

        Computation
            The difference between 12 and 26-day moving averages.

        What it signals
            Rising Moving Average Convergence Divergence (MACD) indicates an
            upward price trend and falling MACD indicates a downward price
            trend.

        Reference:
            https://economictimes.indiatimes.com/
        Args:
            plot (bool): If plot bollinger bands
            plot_name(str): Plot with ".png" extension.
            show_plot: True if you plot should be displayed.


        Returns:
            pd.DataFrame: Pandas DataFrame with MACD values

        """
        data = self.df.copy()
        data["EMA_12"] = data[self.price_col].ewm(span=12, adjust=False).mean()
        data["EMA_26"] = data[self.price_col].ewm(span=26, adjust=False).mean()
        data["MACD"] = data["EMA_12"] - data["EMA_26"]
        data.drop(["EMA_12", "EMA_26"], axis=1, inplace=True)
        data = data.dropna()

        if plot:
            fig, ax = plt.subplots(figsize=(14, 9))
            plt.plot(
                data[self.date_col],
                data[self.price_col],
                color="r",
                label="Price",
            )
            plt.plot(data[self.date_col], data["MACD"], color="b", label="MACD")
            plt.legend()
            plt.title("Price and MACD Plot", fontsize=28, color="b")
            plt.xlabel("Time", color="b", fontsize=19)
            plt.ylabel("Price", color="b", fontsize=19)
            plt.savefig(
                self.plot_dir.joinpath(plot_name),  # type: ignore
                bbox_inches="tight",
                facecolor="orange",
            )
            fig.set_facecolor("orange")
            if show_plot:
                plt.show()
        return data

    def get_simple_moving_average(
        self,
        days: Optional[int] = 15,
        plot: Optional[bool] = False,
        plot_name: Optional[str] = "sma.png",
        show_plot: Optional[bool] = False,
    ) -> pd.DataFrame:
        """
        Simple moving average of given days
        Args:
            days (int): Number of days to calculate SMA
            plot (bool): If plot bollinger bands
            plot_name(str): Plot with ".png" extension.
            show_plot: True if you plot should be displayed.

        Returns:
            pd.DataFrame: Pandas DataFrame with SMA values

        """

        data = self.df.copy()
        data = data.sort_values(by=self.date_col).reset_index(drop=True)
        data["SMA"] = data[self.price_col].rolling(days).mean()
        data = data.dropna()
        data = data.sort_values(by=self.date_col, ascending=False).reset_index(
            drop=True
        )
        if plot:
            fig, ax = plt.subplots(figsize=(14, 9))
            plt.plot(
                data[self.date_col],
                data[self.price_col],
                color="r",
                label="Price",
            )
            plt.plot(data[self.date_col], data["SMA"], color="b", label="SMA")
            plt.legend()
            plt.title("Price and SMA Plot", fontsize=28, color="b")
            plt.xlabel("Time", color="b", fontsize=19)
            plt.ylabel("Price", color="b", fontsize=19)
            plt.savefig(
                self.plot_dir.joinpath(plot_name),  # type: ignore
                bbox_inches="tight",
                facecolor="orange",
            )
            fig.set_facecolor("orange")
            if show_plot:
                plt.show()
        return data

    def get_exponential_moving_average(
        self,
        periods: List[int] = [20],
        plot: Optional[bool] = False,
        plot_name: Optional[str] = "ema.png",
        show_plot: Optional[bool] = False,
    ):
        """
        The EMA is a moving average that places a greater weight and
        significance on the most recent data points. Like all moving averages,
        this technical indicator is used to produce buy and sell signals based
        on crossovers and divergences from the historical average.

        Traders often use several different EMA days, for instance, 20-day,
        30-day, 90-day, and 200-day moving averages.
        Reference:
            https://www.investopedia.com/
        Args:
            periods (list): List of period to calculate EMA
            days (int): Number of days to calculate SMA
            plot (bool): If plot bollinger bands
            plot_name(str): Plot with ".png" extension.
            show_plot: True if you plot should be displayed.

        Returns:
            pd.DataFrame: Pandas DataFrame with EMA values
        """
        data = self.df.copy()
        for period in periods:
            data["EMA_{}".format(period)] = (
                data[self.price_col].ewm(span=period, adjust=False).mean()
            )
        if plot is True:
            fig, ax = plt.subplots(figsize=(14, 9))
            plt.plot(
                data[self.date_col],
                data[self.price_col],
                color="r",
                label="Price",
            )
            for period in periods:
                plt.plot(
                    data[self.date_col],
                    data["EMA_{}".format(period)],
                    label="EMA_{}".format(period),
                )
            plt.legend()
            plt.title("Price and EMA Plot", fontsize=28, color="b")
            plt.xlabel("Time", color="b", fontsize=19)
            plt.ylabel("Price/EMA", color="b", fontsize=19)
            plt.savefig(
                self.plot_dir.joinpath(plot_name),  # type: ignore
                bbox_inches="tight",
                facecolor="orange",
            )
            fig.set_facecolor("orange")
            if show_plot is True:
                plt.show()
        return data
